File: core/src/de/project/ice/screens/MessageScreen.java
package de.project.ice.screens;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.viewport.ScreenViewport;
import de.project.ice.IceGame;
import de.project.ice.utils.ColorDrawable;
import de.project.ice.utils.DelegatingBlockingInputProcessor;
import org.jetbrains.annotations.NotNull;


public class MessageScreen extends BaseScreenAdapter {
    @NotNull private Array<String> messages;

    @NotNull
    private Stage stage;
    @NotNull
    private Skin skin;
    @NotNull
    private Table root;
    @NotNull
    private Label messageLabel;

    public MessageScreen(@NotNull IceGame game, String... messages) {
        super(game);

        this.messages = new Array<String>(messages);
        this.messages.reverse();

        stage = new Stage();
        stage.setViewport(new ScreenViewport());

        skin = new Skin(Gdx.files.internal("ui/skin.json"));

        root = new Table(skin);
        stage.addActor(root);

        root.setFillParent(true);
        root.setBackground(new ColorDrawable(new Color(0f, 0f, 0f, 0.4f), true));

        root.row().expand();
        root.add("");

        messageLabel = new Label(null, skin);
        root.row().expandX();
        root.add(messageLabel);

        root.row().expand();
        root.add("");

        inputProcessor = new DelegatingBlockingInputProcessor(stage) {
            @Override
            public boolean touchDown(int screenX, int screenY, int pointer, int button) {
                showNextMessage();
                return true;
            }
        };

        showNextMessage();
    }

    private void showNextMessage() {
        if (messages.size == 0) {
            game.removeScreen(this);
            return;
        }

        messageLabel.setText(messages.pop());
    }

    @Override
    public int getPriority () {
        return 500;
    }

    public void resize (int width, int height) {
        stage.getViewport().update(width, height, true);
    }

    @Override
    public void update(float delta) {
        stage.act(delta);
    }

    public void render () {
        stage.getViewport().apply();
        stage.draw();
    }

    public void dispose() {
        stage.dispose();
    }
}

File: core/src/de/project/ice/screens/MessageScreen.kt
package de.project.ice.screens

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.viewport.ScreenViewport
import de.project.ice.IceGame
import de.project.ice.utils.*

class MessageScreen(game: IceGame, vararg messages: String, private val timeout: Float = 2.0f) : BaseScreenAdapter(game) {
    private val messages = Array(messages).apply { reverse() }

    private val stage = Stage()
    private val skin = DefaultSkin
    private val root = Table(skin)
    private val messageLabel: Label

    private var remainingTime = timeout + FADE_TIME

    init {
        stage.viewport = ScreenViewport()

        stage.addActor(root)

        root.setFillParent(true)
        root.bottom()

        val background = object: Image(RoundedRectangleDrawable(Color(0f, 0f, 0f, 0.4f), true, 5f)) {
            override fun draw(batch: Batch?, parentAlpha: Float) {
                (drawable as? RoundedRectangleDrawable)?.alpha = parentAlpha
                super.draw(batch, parentAlpha)
            }
        }

        messageLabel = Label(null, skin, "message")

        val stack = Stack(background, Container(messageLabel).pad(5f))
        root.add(Container(stack).padBottom(200f))

        showNextMessage()
    }

    fun addMessages(vararg messages: String) {
        messages.forEach { this.messages.insert(0, it) }
    }

    private fun showNextMessage() {
        if (messages.size == 0) {
            game.removeScreen(this)
            return
        }

        messageLabel.setText(messages.pop())

        stage.addAction(Actions.sequence(Actions.alpha(0f), Actions.fadeIn(FADE_TIME)))
    }

    override val priority: Int
        get() = 90

    override fun resize(width: Int, height: Int) {
        stage.viewport.update(width, height, true)
    }

    override fun update(delta: Float) {
        stage.act(delta)
        remainingTime -= delta
        if (remainingTime <= -FADE_TIME) {
            showNextMessage()
            remainingTime = timeout + FADE_TIME
        } else if (remainingTime <= 0) {
            stage.addAction(Actions.fadeOut(FADE_TIME))
        }
    }

    override fun render() {
        stage.viewport.apply()
        stage.draw()
    }

    override fun dispose() {
        stage.dispose()
    }

    companion object {
        val FADE_TIME = 0.3f
    }
}
